In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn import metrics
from sklearn.preprocessing import LabelEncoder
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split
import json

In [2]:
train_df = pd.read_csv("train.csv")
train_df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

from above table we conclude that:

    1) 38.38% of passengers survived
    
    2) There are some missing values in the 'Age' column and that make about 20 percent of our data threrfore rows with Nan as Age are not erasable.

we need to check for missing values in the non-numeric collumns:

In [3]:
train_df.isnull().sum().sort_values(ascending=False)/train_df.shape[0]

Cabin          0.771044
Age            0.198653
Embarked       0.002245
PassengerId    0.000000
Survived       0.000000
Pclass         0.000000
Name           0.000000
Sex            0.000000
SibSp          0.000000
Parch          0.000000
Ticket         0.000000
Fare           0.000000
dtype: float64

about 77 percent of Cabin and 20 percent of age is missing. we will drop the Cabin and Age column and drop the rows with missing Embraked value.

In [4]:
train_df.drop(columns=['PassengerId', 'Name', 'Ticket', 'Cabin', 'Age'], inplace=True)
train_df.dropna(subset=['Embarked'])

Survived  Pclass     Sex  SibSp  Parch     Fare Embarked
0           0       3    male      1      0   7.2500        S
1           1       1  female      1      0  71.2833        C
2           1       3  female      0      0   7.9250        S
3           1       1  female      1      0  53.1000        S
4           0       3    male      0      0   8.0500        S
..        ...     ...     ...    ...    ...      ...      ...
886         0       2    male      0      0  13.0000        S
887         1       1  female      0      0  30.0000        S
888         0       3  female      1      2  23.4500        S
889         1       1    male      0      0  30.0000        C
890         0       3    male      0      0   7.7500        Q

[889 rows x 7 columns]

In [5]:
train_df['Sex'] = LabelEncoder().fit_transform(train_df['Sex'].values)
train_df['Embarked'] = LabelEncoder().fit_transform(train_df['Embarked'].values)

In [6]:
X = train_df.iloc[:,1:7]
Y = train_df['Survived']

X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.5, random_state=5)

In [7]:
test_df = pd.read_csv('test.csv').drop(columns=['PassengerId', 'Name', 'Ticket', 'Cabin', 'Age'])
test_df['Sex'] = LabelEncoder().fit_transform(test_df['Sex'].values)
test_df['Embarked'] = LabelEncoder().fit_transform(test_df['Embarked'].values)

In [8]:
test_df['Fare'].fillna(test_df['Fare'].mean, inplace=True)

In [9]:
K = range(1, 20) #K is defenitly not optimized :)
scores = []
for k in K:
    knn = KNeighborsClassifier(n_neighbors=k).fit(X_train, Y_train)
    pred = knn.predict(X_test)
    scores.append(metrics.accuracy_score(Y_test, pred))
print(scores)

[0.7556053811659192, 0.7017937219730942, 0.7488789237668162, 0.6995515695067265, 0.7376681614349776, 0.742152466367713, 0.742152466367713, 0.7309417040358744, 0.742152466367713, 0.7107623318385651, 0.7242152466367713, 0.6995515695067265, 0.7017937219730942, 0.7130044843049327, 0.7130044843049327, 0.7174887892376681, 0.726457399103139, 0.6973094170403588, 0.7130044843049327]


Best Accuracy on : K = 1

In [22]:
knn = KNeighborsClassifier(n_neighbors=3).fit(X, Y)
pred = knn.predict(test_df.values)

ValueError: Input contains NaN, infinity or a value too large for dtype('float64').

In [26]:
test_df.dropna(subset=['Fare'])
test_df['Fare'] = pd.to_numeric(test_df['Fare'],errors = 'coerce')


I didn't have much time to debugg this part
Sorry for any inconvinience

In [27]:
test_df.isnull().sum().sort_values(ascending=False)/test_df.shape[0]

Fare        0.002392
Pclass      0.000000
Sex         0.000000
SibSp       0.000000
Parch       0.000000
Embarked    0.000000
dtype: float64

classification is about predicting a label and regression is about predicting a quantity

Classification alghorithms: 
    
    Logistic Regression
    Naïve Bayes
    Stochastic Gradient Descent
    K-Nearest Neighbours
    Decision Tree
    Random Forest
    Support Vector Machine

Regression algorithms:

    Simple Linear Regression model
    Lasso Regression
    Logistic regression
    Support Vector Machines
    Multivariate Regression algorithm
    Multiple Regression Algorithm

Python has a extensive collection of libraries for AI and easier to read therefore it became popular for Machine Learning

I am absolutly obssesed with subjects like Machine learning, Deep learning and Big Data but I feel lost. I hope this course (If u accept me) would be able to give me a clear path and a better underestanding of industrial application of AI.